examples/air-quality-sensor-app/silabs/src/SensorManager.cpp

```diff
@@ -0,0 +1,176 @@
+    {
+        return AirQualityEnum::kModerate;
+    }
+    else if (value < POOR_THRESHOLD)
+    {
+        return AirQualityEnum::kPoor;
+    }
+    else if (value < VERY_POOR_THRESHOLD)
+    {
+        return AirQualityEnum::kVeryPoor;
+    }
+    else
+    {
+        return AirQualityEnum::kExtremelyPoor;
+    }
+}
+
+CHIP_ERROR SensorManager::Init()
+{
+
+    PlatformMgr().LockChipStack();
+    ChipLogDetail(AppServer, "Int instnace");
+    AirQualitySensorManager::InitInstance(1);
+    ChipLogDetail(AppServer, "Int instnace success");
+    PlatformMgr().UnlockChipStack();
+    // Create cmsisos sw timer for air quality sensor timer.
+    mSensorTimer = osTimerNew(SensorTimerEventHandler, osTimerPeriodic, nullptr, nullptr);
+    if (mSensorTimer == NULL)
+    {
+        ChipLogDetail(AppServer, "mSensorTimer timer create failed");
+        return APP_ERROR_CREATE_TIMER_FAILED;
+    }
+
+#ifdef USE_AIR_QUALITY_SENSOR
+    if (SL_STATUS_OK != AirQualitySensor::Init())
+    {
+        ChipLogDetail(AppServer, "Failed to Init Sensor");
+        return CHIP_ERROR_INTERNAL;
+    }
+#endif
+    // Update Air Quality immediatly at bootup
+    SensorTimerEventHandler(nullptr);
+    // Trigger periodic update
+    uint32_t delayTicks = ((uint64_t) osKernelGetTickFreq() * kSensorTImerPeriodMs) / 1000;
+
+    // Starts or restarts the function timer
+    if (osTimerStart(mSensorTimer, delayTicks))
+    {
+        ChipLogDetail(AppServer, "mSensor Timer start() failed");
+        appError(APP_ERROR_START_TIMER_FAILED);
+    }
+    return CHIP_NO_ERROR;
+}
+
+void SensorManager::SensorTimerEventHandler(void * arg)
+{
+    int32_t air_quality;
+#ifdef USE_AIR_QUALITY_SENSOR
+    if (SL_STATUS_OK != AirQualitySensor::GetAirQuality(air_quality))
+    {
+        ChipLogDetail(AppServer, "Failed to read Air Quality !!!");
+        return;
+    }
+#else
+    // Initialize static variables to keep track of the current index and repetition count
+    static uint8_t nbOfRepetition = 0;
+    static uint8_t simulatedIndex = 0;
+
+    // Ensure the simulatedIndex wraps around the array size to avoid out-of-bounds access
+    simulatedIndex = simulatedIndex % ArraySize(mSimulatedAirQuality);
+    // Retrieve the current air quality value from the simulated data array using the simulatedIndex
+    air_quality = mSimulatedAirQuality[simulatedIndex];
+
+    // Increment the repetition count
+    nbOfRepetition++;
+    // Check if the number of repetitions has reached the threshold to simulate a new reading
+    if (nbOfRepetition >= kSimulatedReadingFrequency)
+    {
+        // Move to the next index for the next simulated reading
+        simulatedIndex++;
+        // Reset the repetition count
+        nbOfRepetition = 0;
+    }
+#endif // USE_AIR_QUALITY_SENSOR
+
+    PlatformMgr().LockChipStack();
+    AirQualitySensorManager::GetInstance()->OnAirQualityChangeHandler(classifyAirQuality(air_quality));
+
+    AppTask::GetAppTask().UpdateAirQualitySensorUI();
```

Comment:
using scheduleWork makes it so you don't need to lock. Instead, it places the function call on the chip task queue. Since it will be run on the chip task context it can access the datamodel directly and don't lock the chip task from running.

It however delays a bit the execution of the function.

Reply:
replacing the lock with schedulework for all use cases.